### ISSUE TITLE
[video_player] Flaky integration test failure on web

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#130147]
While converting web platform tests to LUCI, I'm seeing frequent failures in one test:

```
Failure in method: stay paused when seeking after video completed
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════
The following TestFailure was thrown running a test:
Expected: Duration:<0:00:07.534000>
  Actual: Duration:<0:00:07.544000>

When the exception was thrown, this was the stack:
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 294:49  throw_
packages/matcher/src/expect/prints_matcher.dart.js 433:22                     fail
packages/matcher/src/expect/prints_matcher.dart.js 430:12                     _expect
packages/matcher/src/expect/prints_matcher.dart.js 365:12                     expect$
packages/flutter_test/src/test_text_input_key_handler.dart.js 7284:12         expect$
video_player_test.dart.js 257:23                                              <fn>
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 45:50            <fn>
...
```

It wasn't happening in Cirrus, but I could repro it locally in 2/4 runs. Either Cirrus for some reason didn't trip the flake, or the flake only happens in newer versions of Chrome (I believe the LUCI config has a newer Chrome version than our last update to the one pulled in our Cirrus tests).

I'm going to mark this as flaky to unblock the migration.

**Comment by stuartmorgan** [comment]
@ditman FYI

**Comment by ditman** [comment]
I'll get this test deflaked ASAP. I remember there was some other time-dependent code in the test that would flake more (this seems... parsing video metadata?), but I'll tackle this soon.

**Comment by ditman** [comment]
This might be a bug/race condition on the web implementation (?).

At first I thought that "10ms" was too low resolution for a web browser to do anything meaningful with, so I bumped the rewind up to "500ms", but that didn't help; in some cases it seems that:

* the rewind is not applied (hence why the play head remains at the end)
* the video actually plays from the beginning

I need to step through the test case manually and see what's actually happening.

**Comment by ditman** [comment]
TL;DR: I've looked at the player code and the test a little bit, and there are a few problems with the plugin (that we already know about; my upcoming rant is related to: #90080).

----

The plugin **attempts to keep track of the play head position [with a Timer that only refreshes every 500ms](https://github.com/flutter/packages/blob/main/packages/video_player/video_player/lib/video_player.dart#L571-L593)**, but the test is operating in much [shorter increments](https://github.com/flutter/packages/blob/main/packages/video_player/video_player/example/integration_test/video_player_test.dart#L131-L136). The web might not fully respect those timeouts [for many reasons](https://developer.mozilla.org/en-US/docs/Web/API/setTimeout#reasons_for_delays_longer_than_specified).

There's also a couple of `seekTo` calls liberally applied to the code to manipulate the play head position in certain moments, like "when the video ends, jump to the last frame":

* https://github.com/flutter/packages/blob/main/packages/video_player/video_player/lib/video_player.dart#L463-L469

Or "when the user clicks play, if you're in the last frame, rewind to 0":

* https://github.com/flutter/packages/blob/main/packages/video_player/video_player/lib/video_player.dart#L534-L542

The `seekTo` calls above **get sent to the native implementation**, and in the web (at least, I haven't looked at how the defensive is the code in the mobile implementations) causes trouble.

_(It doesn't help that the web implementation has a bug where the `autoplay` attribute, even though we try [very](https://github.com/flutter/packages/blob/main/packages/video_player/video_player_web/lib/src/video_player.dart#L63) [hard](https://github.com/flutter/packages/blob/main/packages/video_player/video_player_web/lib/src/video_player.dart#L69-L70) to make it `false` is actually set to `true` (because it's an [HTML Boolean attribute](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML)). This causes that when the video reaches the end, and is seeked, some times the `play` event gets fired again. This probably caused the "seek to zero" behavior that I was seeing above.)_

In any case, I think a proper solution to this is to **send the progress updates from the native implementation to the plugin, simplify the event handling coming from the native implementation, and stop using `seekTo` "internally".** The Flutter side of the plugin should be super simple, we just need to choose [what events we care about](https://html.spec.whatwg.org/multipage/media.html#mediaevents).


As for why this started failing with us updating the browser, _**and I haven't verified this**_, but [this change](https://bugs.chromium.org/p/chromium/issues/detail?id=1408563) looks _sus_ (and somewhat related to the failing test) ⁉️

**Comment by ditman** [comment]
After [this](https://github.com/flutter/packages/pull/4961) lands, we might try to re-enable the flaky test. I've ran it with the package overridden locally, and it has passed several (8) consecutive attempts.

**Comment by stuartmorgan** [comment]
FWIW, revamping how video_player works internally so it's less "two sources of truth fighting for dominance" is something I am hoping will be a 2024H1 project.

**Comment by ditman** [comment]
I think we're going to be able to close this with only the fix to the `autoplay` issue on the web, so I guess we're going to be fine until then (but yeah the added complexity is super frustrating when debugging!)

**Comment by ditman** [comment]
Haven't had time to work on this one anymore. I suspect this is hard to fix because we're trying to keep in sync the state of the plugin with the internal state of the player, buuut I haven't rewritten this to verify, so: **don't know**.
